internal/stacks/stackruntime/internal/stackeval/stubs/unknown.go

```diff
@@ -0,0 +1,191 @@
+
+package stubs
+
+import (
+	"github.com/zclconf/go-cty/cty"
+
+	"github.com/hashicorp/terraform/internal/providers"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+)
+
+var _ providers.Interface = (*unknownProvider)(nil)
+
+// unknownProvider is a stub provider that represents a provider that is
+// unknown to the current Terraform configuration. This is used when a reference
+// to a provider is unknown, or the provider itself has unknown instances.
+//
+// This provider wraps an unconfigured provider client, which is used to handle
+// offline functionality.
+//
+// TODO: We can return more specific values than cty.DynamicVal.
+type unknownProvider struct {
+	unconfiguredClient providers.Interface
+}
+
+func UnknownProvider(unconfiguredClient providers.Interface) providers.Interface {
+	return &unknownProvider{
+		unconfiguredClient: unconfiguredClient,
+	}
+}
+
+func (u *unknownProvider) GetProviderSchema() providers.GetProviderSchemaResponse {
+	// This is offline functionality, so we can hand it off to the unconfigured
+	// client.
+	return u.unconfiguredClient.GetProviderSchema()
+}
+
+func (u *unknownProvider) ValidateProviderConfig(request providers.ValidateProviderConfigRequest) providers.ValidateProviderConfigResponse {
+	// This is offline functionality, so we can hand it off to the unconfigured
+	// client.
+	return u.unconfiguredClient.ValidateProviderConfig(request)
+}
+
+func (u *unknownProvider) ValidateResourceConfig(request providers.ValidateResourceConfigRequest) providers.ValidateResourceConfigResponse {
+	// This is offline functionality, so we can hand it off to the unconfigured
+	// client.
+	return u.unconfiguredClient.ValidateResourceConfig(request)
+}
+
+func (u *unknownProvider) ValidateDataResourceConfig(request providers.ValidateDataResourceConfigRequest) providers.ValidateDataResourceConfigResponse {
+	// This is offline functionality, so we can hand it off to the unconfigured
+	// client.
+	return u.unconfiguredClient.ValidateDataResourceConfig(request)
+}
+
+func (u *unknownProvider) UpgradeResourceState(request providers.UpgradeResourceStateRequest) providers.UpgradeResourceStateResponse {
+	// This is offline functionality, so we can hand it off to the unconfigured
+	// client.
+	return u.unconfiguredClient.UpgradeResourceState(request)
+}
+
+func (u *unknownProvider) ConfigureProvider(request providers.ConfigureProviderRequest) providers.ConfigureProviderResponse {
+	// This shouldn't be called, we don't configure an unknown provider within
+	// stacks and Terraform Core shouldn't call this method.
+	panic("attempted to configure an unknown provider")
+}
+
+func (u *unknownProvider) Stop() error {
+	// the underlying unconfiguredClient is managed elsewhere.
+	return nil
+}
+
+func (u *unknownProvider) ReadResource(request providers.ReadResourceRequest) providers.ReadResourceResponse {
+	if request.ClientCapabilities.DeferralAllowed {
+		return providers.ReadResourceResponse{
+			NewState: cty.DynamicVal,
```

Comment:
I think I remember we said at some point that we expect whatever the provider returns to be an unknown or dynamic value in the shape of the requested resource. Not sure if there are explicit expectations tied to this or if a step changes this in-between here and where we call these functions in the core graph. I guess the test case would have caught it if there were a problem 🤔

Reply:
Yeah, that's what the TODO was for! Done!